fvm/evm/offchain/blocks/block_context.go

```diff
@@ -0,0 +1,109 @@
+package blocks
+
+import (
+	gethCommon "github.com/onflow/go-ethereum/common"
+	"github.com/onflow/go-ethereum/eth/tracers"
+
+	"github.com/onflow/flow-go/fvm/evm/types"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// NewBlockContext creates a new block context for the given chain ID and height.
+// This is for use in offchain re-execution of transactions.
+// It includes special casing for some historical block heights:
+//   - On Mainnet and Testnet the block hash list was stuck in a loop of 256 block hashes until fixed.
+//     https://github.com/onflow/flow-go/issues/6552
+//   - The coinbase address was different on testnet until https://github.com/onflow/flow-evm-gateway/pull/491.
+func NewBlockContext(
+	chainID flow.ChainID,
+	height uint64,
+	timestamp uint64,
+	getHashByHeight func(uint64) gethCommon.Hash,
+	prevRandao gethCommon.Hash,
+	tracer *tracers.Tracer,
+) (types.BlockContext, error) {
+
+	// coinbase address fix
+	miner := types.CoinbaseAddress
+	if chainID == flow.Testnet && height < coinbaseAddressChangeEVMHeightTestnet {
+		miner = oldCoinbaseAddressTestnet
+	}
+
+	return types.BlockContext{
+		ChainID:                types.EVMChainIDFromFlowChainID(chainID),
+		BlockNumber:            height,
+		BlockTimestamp:         timestamp,
+		DirectCallBaseGasUsage: types.DefaultDirectCallBaseGasUsage,
+		DirectCallGasPrice:     types.DefaultDirectCallGasPrice,
+		GasFeeCollector:        miner,
+		GetHashFunc: func(n uint64) gethCommon.Hash {
+			// For block heights greater than or equal to the current,
+			// return an empty block hash.
+			if n >= height {
+				return gethCommon.Hash{}
+			}
+			// If the given block height, is more than 256 blocks
+			// in the past, return an empty block hash.
+			if height-n > 256 {
+				return gethCommon.Hash{}
+			}
+
+			// For testnet & mainnet, we fetch the block hash from the hard-coded
+			// array of hashes.
+			if chainID == flow.Mainnet && height < blockHashListFixHCUEVMHeightMainnet {
```

Comment:
We need to add this logic to BasicProvider.OnBlockExecuted to call PushBlockHash with the hard-coded array of hashes. 

So that [my backward compatibility test](https://github.com/onflow/flow-go/pull/6749) will be able to save with the hard-coded hashes, and verify the trie update matches. If it matches, then we know these hard-coded configs are correct and can be used by gateway.

Reply:
```suggestion
			if chainID == flow.Mainnet && n < blockHashListFixHCUEVMHeightMainnet {
```

It should be `n` right? height is for the current block

Reply:
I did some refactor in my branch for your reference

https://github.com/onflow/flow-go/commit/2fd43ac17827f43dd5a6ddc3380b976fbe7262af

Reply:
I added a method similar to what you had in your branch